src/Utils/Snapping.vala

```diff
@@ -0,0 +1,296 @@
+    }
+
+    // Grid snaps found for a given selection and set of candidates
+    public struct SnapGrid {
+        public bool is_empty () {
+            return (v_snaps.size == 0 && h_snaps.size == 0);
+        }
+
+        public Gee.HashMap<int, SnapMeta> v_snaps;
+        public Gee.HashMap<int, SnapMeta> h_snaps;
+    }
+
+    // Type of match that was found for a snap
+    public enum MatchType {
+        NONE=-1,    //< No match was found
+        FUZZY,      //< A match was found, but requires an offset
+        EXACT,      //< An exact match was found, no offset is necessary
+    }
+
+    // Information used to define a match for a given selection.
+    // An instance of this class corresponds to a single direction (vertical or horizontal).
+    public struct SnapMatch {
+        // Returns true if a match was found
+        public bool snap_found () {
+            return type != MatchType.NONE;
+        }
+
+        // Returns the offset necessary to bring the reference position to the selected items
+        public int snap_offset () {
+            if (snap_found ()) {
+                return snap_position + polarity_offset - reference_position;
+            }
+            return 0;
+        }
+
+        public MatchType type;
+        public int snap_position; //< Position of matched snap
+        public int polarity_offset; //< Offset incurred by polarity properties. For now it is always zero
+        public int reference_position; //< Position relative to the selection used as a reference
+        public Gee.HashMap<int, int> exact_matches; //< map of exact matches and their polarities
+    }
+
+    // Couple of MatchData used for convenience
+    public struct SnapMatchData {
+        public bool snap_found () {
+            return h_data.snap_found () || v_data.snap_found ();
+        }
+
+        SnapMatch h_data;
+        SnapMatch v_data;
+    }
+
+    /// Returns a sensitivity adjusted to the given canvas scale
+    public static double adjusted_sensitivity (double canvas_scale) {
+        // Limit the sensitivity. This seems like a sensible default for now
+        if (canvas_scale > SENSITIVITY) {
+            return 1.0;
+        }
+
+        // beyond 0.002, the snapping breaks down. Arguably, it does before.
+        return SENSITIVITY / double.max(0.002, canvas_scale);
+    }
+
+    // Generates a snap grid from a canvas.
+    public static SnapGrid snap_grid_from_canvas (
+        Goo.Canvas canvas,
+        List<Lib.Items.CanvasItem> selection,
+        int sensitivity
+    ) {
+        List<weak Goo.CanvasItem> vertical_candidates = null;
+        List<weak Goo.CanvasItem> horizontal_candidates = null;
+
+        Goo.CanvasBounds vertical_filter = {0, 0, 0, 0};
+        Goo.CanvasBounds horizontal_filter = {0, 0, 0, 0};
+
+        foreach (var item in selection) {
+          horizontal_filter.x1 = item.bounds.x1 - sensitivity;
+          horizontal_filter.x2 = item.bounds.x2 + sensitivity;
```

Comment:
Let's avoid using the `item.bounds` as those change for masked groups, artboards, or based on the item's border and don't reflect the correct coordiantes.
For `bounds.x1` and `y1`, you can use the `item.transform.x` and `item.transform.y`.

For x2 and y2, it's tricky.
I think we should create those getters in the `Transform` component to return the correct bounds.